text/separate-compilation.md

```diff
@@ -0,0 +1,229 @@
+- Title: Separate Compilation in Coq
+
+- Drivers: David Swasey, Paolo Giarrusso, Gregory Malecha
+
+# Summary
+
+Coq provides a module system that can be used explicitly through commands such
+as `Module` and `Module Type`. These can be quite heavyweight in many instances,
+and have some limitations when it comes to separately compiling files and
+building generic libraries.
+To address these problems, and support information hiding and Cardelli-style separate
+compilation, in this CEP we propose a notion of Coq interface files, which we
+will call a `.vi` file, and which are inspired by OCaml's `.mli` files.
+Intuitively, a Coq interface file called `module.vi` defines the public
+interface for `module.v`. The `module.vi` interface shall enable developing and
+typechecking clients, even before `module.v` has been implemented.
+
+If both `module.vi` and `module.v` are present, `module.vi` shall act as an
+opaque ascription for the top-level module defined by `module.v`; this opaque
+ascription ensures that clients that typecheck against `module.vi` shall still
+typecheck against the combination of `module.vi` and `module.v`, regardless of
+the implementation details of `module.v`, including any non-logical side effects
+such as hints (but excluding [universes](#universes)). Instead, without
+interfaces, adding implementations is sufficient to break clients, and changing
+implementations can break clients again, hindering modular development.
+
+This has a few advantages compared to Coq's state of the art:
+- It enables separate development: after agreeing on `module.vi`, `module.v` and
+  its clients can be developed independently. To ensure this, unlike today's
+  opaque ascription, `.vi` files can even hide side effects due to `Require`.
+- It reduces compile-time dependencies and improves compile times, even compared
+  to today's `vos` builds (initial builds can be more parallel, and incremental
+  builds need to recompile fewer files).
+
+NOTE: For concreteness, in this CEP we use the `.vi` extension for for interface
+source files, and the `.vio` extension for interface object files. However,
+`.vio` files should not be confused with `.vio` files produced by `-quick`
+builds: to avoid confusion, we could choose other file extensions or remove
+`-quick` builds entirely.
+
+# Proposed Semantics
+
+In this section, we describe the meaning of our Coq extension: instead of
+suggesting an implementation strategy, we sketch a semantics as a
+source-to-source transformation, to be taken as an informal and _approximate_
+specification. However, this is an approximate specification, because the exact
+semantics cannot be expressed via source-to-source transformation.
```

Comment:
I would much prefer a version where this can be fully implemented as a source-to-source translation (in part because that makes things easier for the bug minimizer).  I think it also solves some design questions without making things significantly worse for the user.

I propose the following:
1. Interfaces are interpreted as module types
2. Compiling a .v file with a corresponding .vi acts as if the module type were ascribed to it (except possibly `Require`s, which may be considered to be above the module type ascription as far as global side effects go)
3. At `Require` time there should be three flavors: a file can `Require` just the compiled .vi file, it can `Require` both the compiled .vi file and the compiled .v file, or it can `Require` just a compiled .v file.  The last should be an error if a .vi file (compiled or source) exists alongside the .v/.vo file.  Requiring both the .v(o) and compiled .vi should be like Requiring the compiled .vi and then attaching bodies to the constants for `Print Assumptions`, etc (in the same way it would happen for modules opaquely ascribed with a type), and also including the Requires of the .v file.

This pushes all the issues with side effects of require, universe constraints, ability to use Extraction, etc, onto the user side in a systematic way, by putting linking under the user's control (linking corresponds to requiring the compiled .v file).

Reply:
> except possibly Requires, which may be considered to be above the module type ascription as far as global side effects go

To actually achieve separate compilation, the `Require`s would have to be inside the module and be hidden by the opaque ascription. Today this gives a warning, but in the Coq call it wasn't clear why, but until that's resolved I consider a source-to-source translation impossible _today_ :-).

Reply:
This is what I tried to explain (very badly) by saying that an interface should morally be
```coq
Require stuff.
Module Type this.
Import some stuff.
...
End this.
```

There are a few "actions" performed by require which don't play well with modules, which are usually seen as a linguistic construct to limit the scope of thing.

Ideally, if one does not import a module, he does not get any effect. But require, today, leaks stuff, like global hints. So placing it inside a (sub) module does not do what you may think. Also, there is special code hoisting some actions at the top level, like loading ML plugins and co which is ugly and has no real use case.

Finally, it would be nice to have a header in each file which lets one compute the dependencies of the file easily. If requires are only allowed in there, then we kind of have this header. If we allow them to be anywhere, well... Similarly, require is slow, if they are all at the beginning of the file, a UI pays the cost once upfront, and then it is good to go. If they are interleaved with other sentences, well, you may need to backtrack over that slow line way more often.

To sum up, it is both an "hygienic" and would allow some code simplification if all requires were on top.

Reply:
Generally (though this is a fairly purist point of view), I have been thinking of .v files as corresponding to *functors* from their `Require`d dependencies. Currently, Coq does not provide a mechanism to ascribe a type to these modules, so there is no means to abstract. With this approach, you can change what the module type being abstracted over is which provides more flexibility to the implementer of the interface. If I follow this mental model, then the second flavor of `Require` that you describe above means that you can "abstract over" the "natural" module type (the one that does not hide anything, including the body of opaque definitions); however, it becomes a bit odd because to typecheck those definitions you might need the body of other constants which could introduce problems.